### PR TITLE
[#797] Update linear interpolation of percentiles

### LINKED ISSUE #797: Feature: Have other linear interpolation for percentiles
* xclim version: 0.28
* Python version: 3.x
* Operating System: Mac os

### Description
Follow up to the discussion with @huard.
In other climate indices libraries such as climdex and icclim, the percentile interpolation method used is the 8th method of[ Hyndman and Fan](https://www.amherst.edu/media/view/129116/original/Sample+Quantiles.pdf). Both library have their own C/C++ implementation.
In Xclim, the default interpolation of numpy is used, which correspond to the 7th method of Hyndman and Fan.
However **numpy does not** implement the 8th method (see https://github.com/numpy/numpy/issues/10736#issuecomment-897851315).

It would be great if xclim:
- Use the Scipy [mquantiles method](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mstats.mquantiles.html?highlight=mquantiles) instead of `np.percentile`
- Have a parameter to choose the interpolation in `calendar.percentile_doy`

As usual, I can work on this issue.

### PR DESCRIPTION
**Pull request by bzah**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #797 
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [ ] `bumpversion (major / minor / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)
- [ ] The relevant author information has been added to `.zenodo.json`

### What kind of change does this PR introduce?
Update `_calc_perc` and `percentile_doy` functions to allow more linear interpolation methods.

### Does this PR introduce a breaking change?
I'm not sure how impactful it is but `_calc_perc` now return a numpy masked array.

### CODE REVIEW

**Comment by bzah** [comment]
A remark about the failing tests here.
I think it is quite counter productive to have two places where the same things get tested.

For example, the test from `test_indices.TestTGXN10p.test_tg10p_simple` and  the one from `test_temperature.TestT10p.test_tg10p_simple` are both testing that tg10p is properly computed.
IMO, the test of `atmo.tg10p` should only ensure that the `Temp` object is properly created with the expected properties.

**Comment by aulemahal** [comment]
@bzah good point! Indicators need to be tested, but there's no point to repeat the same tests. Recently, I think we went with this guideline:
- Indice tests (`test_indices.py`) use idealized data and really test the mathematical computation. (like edgecases, presence or not of missing data).
- Indicator tests (`test_{var kind}.py` use "real" data (usually `atmos_ds`, the ERA5 extraction) and test minimally for values, but add tests for attributes and stuff like that.

**Comment by bzah** [comment]
I'm getting really good results with my implementation, I still have to write some unit tests, some documentation and clean up a bit but the PR will be reviewable soon.

**Comment by bzah** [comment]
I ended up completely rewriting numpy functions ::percentile and ::nanpercentile

It would be really appreciated if you can give me your insights on this implementation.
However I will also try to submit it to numpy because:
 - both function show better performances, especially for nanpercentiles which is up to x50 in some cases.
 - they are, IMO, much more readable than current implementations.
 - they add some error correction using floating point epsilon (based on the R implementation)

So I understand if you prefer to wait for the PR on numpy before commenting.

I also wrote a script to compare performance an assert the non regression (up to 14th decimal):
https://gist.github.com/bzah/2a84d050b8a1aed1b40a2ed1526e1f12
In the gist comment you can find a report of some test runs.

**Review by huard — COMMENTED** [review]
I suggest we prep this PR for merging, irrespective of what NumPy is doing. There are impressive speed-ups for nans that are worth having ASAP. 

I suggest adding the reference to the Hyndman paper in the docs, and add some guidance to users on which values should be chosen for which occasion. 

`calc_percentiles` loops over `_quantile` for each percentile. Is there any way to avoid that ? Once `arr` is sorted, I would think finding other percentiles is quick. I'm hopeful the slow-down you measure can be reduced by passing the array of percentiles to `_quantile`. I suspect numpy folks will make the same comment.

**Comment by bzah** [comment]
@huard 

> calc_percentiles loops over _quantile for each percentile. Is there any way to avoid that ? Once arr is sorted, I would think finding other percentiles is quick. I'm hopeful the slow-down you measure can be reduced by passing the array of percentiles to _quantile. I suspect numpy folks will make the same comment.

I was expecting discussion on this .
Yes, I agree with you it would improve the performances when there are multiple quantiles, and it's more numpy style anyway. However, I find it much easier to manipulate a scalar to understand the code and the performance are not even that bad even with 100 quantiles.

I will try to see how I can make it work with a np.array.

About the doc, once in numpy the API will probably be slightly different. 
Because I'm lazy, I wanted to write docs only on numpy but I can start it here and reuse it I suppose.

**Comment by bzah** [comment]
@huard  I did something that work but I'm pretty sure it is not optimal.
I tested with a smallish array (100, 180, 30), iterated 10 times and on 1000 percentiles (from 0.1 to 99.9, step 0.1).
I have these results:
```
nanpercentile: 19.904847238 secs
NEW nan_calc_percentiles: 22.494946018 secs
OLD nan_calc_percentiles: 50.227164944 secs
```
I suppose I'm not doing it properly, it should perform better than nanpercentile.
I'm open to suggestion if someone has a better idea.

I'm coming back Tuesday, I will try to improve it then.

**Comment by bzah** [comment]
About the test failure:
The failing test is in *test_adjustment.py*, I'm not 100% sure but I don't think I broke it with this feature.
Moreover the test works fine locally, is there something I'm missing ?

The error is`AttributeError: 'DataArray' object has no attribute 'units'` on *TestDQM::test_mon_U[True-*-pr]  *

A part from this error, I think the feature is ready for review.

File: HISTORY.rst
Comment (bzah) [inline HISTORY.rst:206]: oops

**Comment by aulemahal** [comment]
> The failing test is in test_adjustment.py

Not your fault! This test shows a seemingly random bug in dask...

**Comment by bzah** [comment]
FYI I created the draft PR on numpy: https://github.com/numpy/numpy/pull/19857
I did quite a few changes to fit the needs of numpy API, I don't believe they are necessary in Xclim.

**Comment by bzah** [comment]
I will remove the `fuzz` part, it may causes license issues because it is inspired by R and climdex which have GPL licenses.

File: HISTORY.rst
Comment (huard) [inline HISTORY.rst:26]:
```suggestion
* ``calendar.py::percentile_doy`` now uses the 8th method of Hyndman & Fan for linear interpolation (alpha = beta = 1/3). Previously, the function used Numpy's percentile, which corresponds to the 7th method. This change is motivated by the fact that the 8th is recommended by Hyndman & Fay and it ensures consistency with other climate indices packages (climdex, icclim).
```

File: HISTORY.rst
Comment (huard) [inline HISTORY.rst:25]:
```suggestion
* ``utils.py::nan_calc_percentiles`` now uses a custom algorithm instead of ``numpy::nanpercentiles`` to have more flexibility on the interpolation method. The performance is also improved. 
```

File: HISTORY.rst
Comment (huard) [inline HISTORY.rst:27]:
```suggestion
* ``utils.py::_cal_perc`` is now only a proxy for ``utils.py::nan_calc_percentiles`` with some axis moves.
```
and move to Internal changes section.

File: xclim/core/calendar.py
Comment (huard) [inline xclim/core/calendar.py:455]:
```suggestion
    Different quantile estimators can be used by specifying `alpha` and `beta` according to specifications given by [Hyndman&Fan]_. The default definition corresponds to method 8, which meets multiple desirable statistical properties for sample quantiles. Note that `numpy.percentile` corresponds to method 7, with alpha and beta set to 1.
```

File: xclim/core/calendar.py
Comment (huard) [inline xclim/core/calendar.py:468]:
```suggestion
    alpha: float
        Plotting position parameter.
    beta: float
        Plotting position parameter.
```

File: xclim/core/calendar.py
Comment (huard) [inline xclim/core/calendar.py:475]:
```suggestion
      Nan will be masked
      
    References
    ----------
    .. [Hyndman&Fan] Hyndman, R. J., & Fan, Y. (1996). Sample quantiles in statistical packages. The American Statistician, 50(4), 361-365.  
```

File: xclim/core/utils.py
Comment (huard) [inline xclim/core/utils.py:247]:
```suggestion
    Notes
    -----
```

File: xclim/core/utils.py
Comment (huard) [inline xclim/core/utils.py:302]: Don't you need to call `any` on a specific axis ?

File: xclim/testing/tests/test_precip.py
Comment (huard) [inline xclim/testing/tests/test_precip.py:389]: So there can be quite significant differences in results. Interesting.

File: xclim/testing/tests/test_utils.py
Comment (huard) [inline xclim/testing/tests/test_utils.py:104]:
```suggestion
        res = nan_calc_percentiles(arr, percentiles=[40.0], alpha=1, beta=1)
```
I know these are the default values, but just for clarity.

**Review by huard — APPROVED** [review]
Nice !

File: xclim/core/utils.py
Comment (bzah) [reply xclim/core/utils.py:303]: valid_values_count is derivative from the  valid_values_count which himself comes from a scalar (value_count) and a reduced array (nan_count) so I don't think we need an axis parameter.

However it might be a code smell that the execution flow is hard to follow and it needs some refactoring.

**Comment by bzah** [comment]
I did quite a lot of changes in the numpy PR which greatly simplify the code of _quantile.
However, **the performance improvement has been dropped** because it was not part of the initial issue in numpy.
Thus, the xclim version performs much better but:
1. it does not handle some cases such as:
   - when dtype is Fraction (object) or bool
   - using `out` parameter to mutate an existing array
   - using `keepdims` to reshape the result to the input dimensions
   - using  `overwrite_input` to qvoid copying the input array
   - having other interpolation options
2. It is very hard to read (thus maintain) `_nan_quantile` because I did the refactoring only on numpy.

My main issue is that I don't know if the performance improvement will be merged or not on numpy.
And I think the boost is too good to be left behind, especially when I think about bootstrapping which call percentile_doy very often.

So, ideally, I would like keep the xclim implementation and refactor it, but to do it only once I'm done with numpy's version to avoid going back and forth between the two.
Is that okay for you ?

**Comment by huard** [comment]
@bzah Is this ready to be merged ?

**Comment by bzah** [comment]
I have not yet done the refactoring because I'm waiting for a final review on the numpy's PR.
If you would like to include this asap, it can be merged as is and I will refactor in a future PR. 
The behavior will not change.

**Review by aulemahal — APPROVED** [review]
I didn't dive in the `nan_calc_percentiles` code. This is great! Clearly, `numpy` is a better place for the function, but xclim can serve as a first home. 

A few nitpicky comments.

File: xclim/core/calendar.py
Comment (aulemahal) [inline xclim/core/calendar.py:454]:
```suggestion
    Different quantile estimators can be used by specifying `alpha` and `beta` according to specifications given by [HyndmanFan]_. The default definition corresponds to method 8, which meets multiple desirable statistical properties for sample quantiles. Note that `numpy.percentile` corresponds to method 7, with alpha and beta set to 1.
```
Sphinx seems to not like the &.

**Comment by huard** [comment]
I suggest to merge now, and open a new refactoring issue later in the future.

**Pull request merged**
